### PR TITLE
Support @optional and @extensibility annotations [19389]

### PR DESCRIPTION
**Pull request by richiware**

### CODE REVIEW

File: src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java [hunk 1]
```diff
@@ -27,53 +27,63 @@
...
+            System.out.println(ColorMessage.error() + msg.toString());
         }
     }
 
-    static private String m_loaderDirectories = "com/eprosima/idl/templates";
-    private Map<String, StringTemplateGroup> m_groups = null;
+    private Map<String, STGroup> m_groups = null;
     private Map<String, List<TemplateExtension>> m_extensions = null;
-    private StringTemplateGroup strackgr_ = null;
+    private STGroup strackgr_ = null;
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java:64]: I suggest removing this variable. It is used in some code that it has been commented out and in some methods that does not seem to be in use (I suggest removing these methods too):

* `addGroup(String, List<TemplateExtension>)`
* `createStringTemplate`

File: src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java [hunk 2]
```diff
@@ -82,28 +92,24 @@ public TemplateManager(String stackTemplateNames, Context ctx, boolean generate_
...
-    {
-        m_loaderDirectories += ":" + paths;
+        */
     }
 
     public void changeCppTypesTemplateGroup(String templateName)
     {
-        TypeCode.cpptypesgr = StringTemplateGroup.loadGroup(templateName, DefaultTemplateLexer.class, null);
+        TypeCode.cpptypesgr = new STGroupFile(templateName, '$', '$');
     }
```
Comment (JLBuenoLopez-eProsima) [inline src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java:108]: This method does not seem to be to in use. Please, remove with the other legacy code.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java [hunk 3]
```diff
@@ -27,53 +27,63 @@
...
+        TypeCode.ctypesgr = new STGroupFile("com/eprosima/idl/templates/CTypes.stg", '$', '$');
+        TypeCode.javatypesgr = new STGroupFile("com/eprosima/idl/templates/JavaTypes.stg", '$', '$');
         TypeCode.ctx = ctx;
 
-        m_groups = new HashMap<String, StringTemplateGroup>();
+        m_groups = new HashMap<String, STGroup>();
         m_extensions = new HashMap<String, List<TemplateExtension>>();
 
         // Load specific template rules.
+        /*
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java]: Done in 3b643a6

File: src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java (hunk 1 above)
Comment (richiware) [reply src/main/java/com/eprosima/idl/generator/manager/TemplateManager.java]: Done in 3b643a6

File: src/main/java/com/eprosima/idl/generator/manager/TemplateGroup.java [hunk 4]
```diff
@@ -74,107 +73,109 @@ public void setAttribute(String attribute, TemplateGroup tg)
...
-            Map.Entry<String, StringTemplate> m = (Map.Entry<String, StringTemplate>)it.next();
+            Map.Entry<String, ST> m = (Map.Entry<String, ST>)it.next();
 
             // Call setAttribute
             Log.printDebug("setting attribute (obj1,obj2,obj3,obj4,obj5) to template group " + m.getKey() + " to " + m.getValue().getName());
-            m.getValue().setAttribute(attribute, obj1, obj2, obj3, obj4, obj5);
+            m.getValue().add(attribute, obj1, obj2, obj3, obj4, obj5);
         }
     }
+    */
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/main/java/com/eprosima/idl/generator/manager/TemplateGroup.java]: Done in 3b643a6

File: src/main/java/com/eprosima/idl/parser/typecode/UnionTypeCode.java [hunk 5]
```diff
@@ -215,6 +218,33 @@ public boolean isIsPlain()
...
+    {
+        int position = 0;
+        List<Map.Entry<Integer, Member>> ret_members = new ArrayList<Map.Entry<Integer,Member>>();
+        AbstractMap.SimpleEntry<Integer, Member> default_member = null;
+
+        for (Member m : getMembers())
+        {
+            if (position == m_defaultindex)
+            {
+                default_member = new AbstractMap.SimpleEntry<>(++position, m);
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/main/java/com/eprosima/idl/parser/typecode/UnionTypeCode.java]: Done in 3b643a6

File: test/idls/optional.idl [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+{
+    unsigned long align;
+    @optional short value[10];
+};
+
+struct InnerStruct //TODO Remove when rebasing
+{
+    short s;
+    long l;
+};
```
[the remark this answers is not in the document]
Comment (richiware) [reply test/idls/optional.idl]: Done in 3b643a6

File: test/idls/coverage_complex.idl [hunk 7]
```diff
@@ -1,94 +0,0 @@
...
-
-struct CompleteStruct
-{
-    MyUnion my_union;
-    MyUnion2 my_union_2;
-};
-
-struct KeyedStruct
-{
-    @Key octet my_key;
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/coverage_complex.idl:92]: This annotation is not being currently tested (after removing these IDL files).

File: test/idls/forward_decls.idl [hunk 8]
```diff
@@ -1,117 +0,0 @@
-union ForwardUnion;
-
-@annotation MyAnnotation
-{
-    uint32 my_value;
-    string name;
-};
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/forward_decls.idl:7]: Annotation definition is not being tested either (and the [documentation](https://fast-dds.docs.eprosima.com/en/latest/fastddsgen/dataTypes/dataTypes.html#annotations) includes this feature as supported)

File: test/idls/forward_decls.idl [hunk 9]
```diff
@@ -1,117 +0,0 @@
...
-    hola::ForwardRecursiveUnboundedSeq rec_unbounded;
-    hola::ForwardRecursiveBoundedSeq rec_bounded;
-};
-
-module hola
-{
-    @MyAnnotation(my_value=55,name="AnnotationName")
-    struct ForwardStruct
-    {
-        @min(5) long mylong_two;
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/forward_decls.idl:38]: Impressive! Testing that the annotation @min is parsed, even if it is not really implemented into the generated code :laughing:

File: test/idls/forward_decls.idl [hunk 10]
```diff
@@ -1,117 +0,0 @@
...
-{
-    long mylong;
-    string<20> mystring;
-    hola::ForwardRecursiveUnboundedSeq rec_unbounded;
-    hola::ForwardRecursiveBoundedSeq rec_bounded;
-};
-
-module hola
-{
-    @MyAnnotation(my_value=55,name="AnnotationName")
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/forward_decls.idl:35]: Annotation application was also tested, though this is not generated into the code per se, so with the current test suite I am not sure if there is an easy way to check it.

File: test/idls/forward_decls.idl [hunk 11]
```diff
@@ -1,117 +0,0 @@
...
-    bitset MyBitset : ParentBitset
-    {
-        bitfield<3> a;
-        bitfield<1> b;
-        bitfield<4>;
-        bitfield<10> c;
-        bitfield<12, short> d, e, f;
-    };
-
-    @bit_bound(8)
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/forward_decls.idl:89]: Another untested annotation

File: test/idls/new_features_4_2.idl [hunk 12]
```diff
@@ -1,84 +0,0 @@
-struct NewAliases
-{
-    int8 int8_;
-    uint8 uint8_;
-    int16 int16_;
-    uint16 uint16_;
-    int32 int32_;
-    uint32 uint32_;
-    int64 int64_;
-    uint64 uint64_;
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/new_features_4_2.idl:10]: This new types are not being tested with the new IDLs.

File: test/idls/include.idl [hunk 13]
```diff
@@ -1,11 +0,0 @@
-#include "basic.idl"
-
-// Issue #534 (Fast-RTPS)
-const octet CONST_DEFINITION_VALUE_0 = 0;
-const octet CONST_DEFINITION_VALUE_1 = 1;
```
Comment (JLBuenoLopez-eProsima) [inline test/idls/include.idl:5]: I am not sure if this regression test has been kept in the new test suite.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Most comments are related to features that were tested in the previous test suite but that now are not tested after removing the IDLs from this repository.

I am also not sure if the extensibility annotation check is implemented correctly. XTypes v1.3 Table 21 specifies that these annotations apply to every Type Declaration. On the other hand, Table 12 specifies the impact of the annotation for the different types. The annotation only affects aggregated types, enumerated types and alias types. I think this implementation has only taken into account the aggregated types.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
After reviewing the XTypes v1.3 specification again, clause 7.3.1.2.1.8 states:

> This annotation may be applied to the definitions of aggregated types. It shall be considered an error for it to be applied to the same type multiple times.

So current implementation that only checks aggregated types for multiple annotation definition seems to be correct.

The other comments have to be implemented in the new [DDS Types Test](https://github.com/eProsima/dds-types-test) repository, so I am approving this PR

**Pull request merged**